avalon/tools/inputs.py

```diff
@@ -0,0 +1,228 @@
+        input.setValue(default)
+        input.setMinimum(min)
+        input.setMaximum(max)
+
+        self.layout.addWidget(input, stretch=True)
+        self.input = input
+
+    def get(self):
+        return self.input.value()
+
+
+def line_edit_float(default=0.0):
+    """Return a QLineEdit widget for inputting float value
+
+    Args:
+        default (float, optional): Default 0.0
+
+    """
+    widget = QtWidgets.QLineEdit()
+    widget.setValidator(QtGui.QDoubleValidator())
+    widget.setText(str(float(default)))
+    return widget
+
+
+class Float(InputBase):
+    """Option box input widget for `float` type value
+
+    Args:
+        name (str): Value entry name, also taken as widget label
+        default (float, optional): Default 0.0
+        help (str, optional): Widget tool tip
+
+    """
+
+    def __init__(self, name, default=0.0, help=None, parent=None):
+        InputBase.__init__(self, name, help, parent)
+
+        input = line_edit_float(default)
+
+        self.layout.addWidget(input)
+        self.input = input
+
+    def get(self):
+        return float(self.input.text())
+
+
+class Double3(InputBase):
+    """Option box input widget for array of three doubles
+
+    Args:
+        name (str): Value entry name, also taken as widget label
+        default (tuple or list, optional): Default (0.0, 0.0, 0.0)
+        help (str, optional): Widget tool tip
+
+    """
+
+    def __init__(self, name, default=None, help=None, parent=None):
+        default = default or (0.0, 0.0, 0.0)
+        assert isinstance(default, (list, tuple)), "Should be list or tuple."
+        assert len(default) == 3, "Should have exact three elements."
+
+        InputBase.__init__(self, name, help, parent)
+
+        input = QtWidgets.QWidget()
+
+        x, y, z = (line_edit_float(v) for v in default)
+
+        layout = QtWidgets.QHBoxLayout(input)
+        layout.addWidget(x)
+        layout.addWidget(y)
+        layout.addWidget(z)
+
+        self.layout.addWidget(input)
+        self.inputs = (x, y, z)
+
+    def get(self):
+        return tuple(float(i.text()) for i in self.inputs)
+
+
+class String(InputBase):
+    """Option box input widget for `str` type value
+
+    Args:
+        name (str): Value entry name, also taken as widget label
+        default (str, optional): Default None
+        placeholder (str, optional): Widget placeholder text, default None
+        help (str, optional): Widget tool tip
+
+    """
+
+    def __init__(self,
+                 name,
+                 default=None,
+                 placeholder=None,
+                 help=None,
+                 parent=None):
+        InputBase.__init__(self, name, help, parent)
+
+        input = QtWidgets.QLineEdit()
+        input.setPlaceholderText(placeholder or "")
+        input.setText(default or "")
+
+        self.layout.addWidget(input)
+        self.input = input
+
+    def get(self):
+        return self.input.text()
+
+
+class GetOne(InputBase):
```

Comment:
Would it be better to name this ComboBox to align with Qt?

Reply:
Hmmm, I am not sure. At the time I thought `ComboBox` wasn't straight forward enough to express *"Select one from the list"*, maybe `Enum` ?

Reply:
Right, I see what you are going for. I thought it was more related to the graphic, cause I was also confused about not using radio buttons instead of combo list.
`GetOne` does seem to stand out as a name. `Enum` might fall better into the mix of naming.

Reply:
Ah yes, it's actually based on what type of value to be inputted and how it will returned.

And yeah, radio buttons could also fit the case of "Select one", so I even thought about implementing both and make an argument to decide which type of widget to use. But I gave up on that because of my laziness :P

Anyway, other potential needs of visualizing key-value pairs into different kinds widgets, can still be added ! See how this go :relaxed:

Reply:
In fact, maybe we could make option dialog to support [`qargparse`](https://github.com/mottosso/qargparse.py), or even drop the `avalon.tool.inputs` ?

Reply:
I just remembered that qargparse was used in Allzpark; it's used for all options in that GUI, most prominently in the Preferences tab was made with, and as a single-file library it's embedded into `vendor/`. We could do something like that for Avalon too; it's worked well so far, and there's lots of room to grow.

- https://github.com/mottosso/allzpark/blob/master/allzpark/dock.py#L947

The only crux is that I'm on other things currently and wouldn't be of much help. So it's up to you to determine whether your time is best spent on a library you've built and fullt understand, or whether you'd be happy diving into and progressing qargparse - something I intend on returning to once I'm back doing pipeline.